File: Frontend/default_abilities.py
from drawClasses import AbilityVisual, EventVisual
from constants import DARK_PINK, STANDARD_RIGHT_CLICK, YELLOW, BLACK, PURPLE, LIGHT_BLUE, PINK, DARK_ORANGE, GREEN, RAGE_CODE, BURN_CODE, \
    ZOMBIE_CODE, CAPITAL_CODE, FREEZE_CODE, POISON_CODE, NUKE_CODE, D_BRIDGE_CODE, BRIDGE_CODE, SPAWN_CODE, \
          GREY, CANNON_CODE, CANNON_SHOT_CODE, STANDARD_LEFT_CLICK
from clickTypeEnum import ClickType


SPAWN_V = AbilityVisual("Spawn", "circle", (None,))
BRIDGE_V = AbilityVisual("Bridge", "triangle", YELLOW, 'A')
D_BRIDGE_V = AbilityVisual("D-Bridge", "circle", YELLOW)
NUKE_V = AbilityVisual("Nuke", "x", BLACK)
POISON_V = AbilityVisual("Poison", "circle", PURPLE)
FREEZE_V = AbilityVisual("Freeze", "triangle", LIGHT_BLUE)
CAPITAL_V = AbilityVisual("Capital", "star", PINK)
ZOMBIE_V = AbilityVisual("Zombie", "square", BLACK)
BURN_V = AbilityVisual("Burn", "square", DARK_ORANGE)
RAGE_V = AbilityVisual("Rage", "cross", GREEN)
CANNON_V = AbilityVisual("Cannon", "cannon", GREY, 'E')

CANNON_SHOT_V = EventVisual("Cannon Shot", DARK_PINK)
STANDARD_LEFT_CLICK_V = EventVisual("Switch", GREY)
STANDARD_RIGHT_CLICK_V = EventVisual("Swap", GREY)

VISUALS = {
    SPAWN_CODE: SPAWN_V,
    BRIDGE_CODE: BRIDGE_V,
    D_BRIDGE_CODE: D_BRIDGE_V,
    NUKE_CODE: NUKE_V,
    POISON_CODE: POISON_V,
    FREEZE_CODE: FREEZE_V,
    CAPITAL_CODE: CAPITAL_V,
    ZOMBIE_CODE: ZOMBIE_V,
    BURN_CODE: BURN_V,
    RAGE_CODE: RAGE_V,
    CANNON_CODE: CANNON_V,

    CANNON_SHOT_CODE: CANNON_SHOT_V,
    STANDARD_LEFT_CLICK: STANDARD_LEFT_CLICK_V,
    STANDARD_RIGHT_CLICK: STANDARD_RIGHT_CLICK_V
}

CLICKS = {
    SPAWN_CODE: (1, ClickType.NODE),
    BRIDGE_CODE: (2, ClickType.NODE),
    D_BRIDGE_CODE: (2, ClickType.NODE),
    NUKE_CODE: (1, ClickType.NODE),
    POISON_CODE: (1, ClickType.NODE),
    FREEZE_CODE: (1, ClickType.EDGE),
    CAPITAL_CODE: (1, ClickType.NODE),
    ZOMBIE_CODE: (1, ClickType.NODE),
    BURN_CODE: (1, ClickType.NODE),
    RAGE_CODE: (0, ClickType.BLANK),
    CANNON_CODE: (2, ClickType.NODE),
}

EVENTS = {
    CANNON_SHOT_CODE: (2, ClickType.NODE),
    STANDARD_LEFT_CLICK: (1, ClickType.EDGE),
    STANDARD_RIGHT_CLICK: (1, ClickType.EDGE)
}

File: TypeScriptFrontend/src/game/slav/default_abilities.ts
import {
    AbilityVisual,
    createAbilityVisual,
    createEventVisual,
    EventVisual,
} from "./immutable_visuals";
import { Colors, KeyCodes, EventCodes } from "./constants";
import { ClickType } from "./enums";

// Create instances of AbilityVisual using the factory function
const SPAWN_V = createAbilityVisual("Spawn", "circle", Colors.GREEN);
const BRIDGE_V = createAbilityVisual("Bridge", "triangle", Colors.YELLOW, "A");
const D_BRIDGE_V = createAbilityVisual("D-Bridge", "circle", Colors.YELLOW);
const NUKE_V = createAbilityVisual("Nuke", "x", Colors.BLACK);
const POISON_V = createAbilityVisual("Poison", "circle", Colors.PURPLE);
const FREEZE_V = createAbilityVisual("Freeze", "triangle", Colors.LIGHT_BLUE);
const CAPITAL_V = createAbilityVisual("Capital", "star", Colors.PINK);
const ZOMBIE_V = createAbilityVisual("Zombie", "square", Colors.BLACK);
const BURN_V = createAbilityVisual("Burn", "square", Colors.DARK_ORANGE);
const RAGE_V = createAbilityVisual("Rage", "cross", Colors.GREEN);
const CANNON_V = createAbilityVisual("Cannon", "cannon", Colors.GREY, "E");
const PUMP_V = createAbilityVisual("Pump", "circle", Colors.GREEN, "U");

// Create instances of EventVisual using the factory function
const CANNON_SHOT_V = createEventVisual("Cannon Shot", Colors.DARK_PINK);
const PUMP_DRAIN_V = createEventVisual("Pump Drain", Colors.YELLOW);
const STANDARD_LEFT_CLICK_V = createEventVisual("Switch", Colors.GREY);
const STANDARD_RIGHT_CLICK_V = createEventVisual("Swap", Colors.GREY);

interface EventVisualParameters {
    [index: string]: EventVisual | AbilityVisual;
}

// Maps for visuals
export const VISUALS: EventVisualParameters = {
    [KeyCodes.SPAWN_CODE]: SPAWN_V,
    [KeyCodes.BRIDGE_CODE]: BRIDGE_V,
    [KeyCodes.D_BRIDGE_CODE]: D_BRIDGE_V,
    [KeyCodes.NUKE_CODE]: NUKE_V,
    [KeyCodes.POISON_CODE]: POISON_V,
    [KeyCodes.FREEZE_CODE]: FREEZE_V,
    [KeyCodes.CAPITAL_CODE]: CAPITAL_V,
    [KeyCodes.ZOMBIE_CODE]: ZOMBIE_V,
    [KeyCodes.BURN_CODE]: BURN_V,
    [KeyCodes.RAGE_CODE]: RAGE_V,
    [KeyCodes.CANNON_CODE]: CANNON_V,
    [KeyCodes.PUMP_CODE]: PUMP_V,

    [EventCodes.CANNON_SHOT_CODE]: CANNON_SHOT_V,
    [EventCodes.PUMP_DRAIN_CODE]: PUMP_DRAIN_V,
    [EventCodes.STANDARD_LEFT_CLICK]: STANDARD_LEFT_CLICK_V,
    [EventCodes.STANDARD_RIGHT_CLICK]: STANDARD_RIGHT_CLICK_V,
};

interface ClickParameters {
    [index: string]: [number, ClickType]; // Adjust the types according to what Event constructor expects
}

// Map for click behaviors
export const CLICKS: ClickParameters = {
    [KeyCodes.SPAWN_CODE]: [1, ClickType.NODE],
    [KeyCodes.BRIDGE_CODE]: [2, ClickType.NODE],
    [KeyCodes.D_BRIDGE_CODE]: [2, ClickType.NODE],
    [KeyCodes.NUKE_CODE]: [1, ClickType.NODE],
    [KeyCodes.POISON_CODE]: [1, ClickType.EDGE],
    [KeyCodes.FREEZE_CODE]: [1, ClickType.EDGE],
    [KeyCodes.CAPITAL_CODE]: [1, ClickType.NODE],
    [KeyCodes.ZOMBIE_CODE]: [1, ClickType.NODE],
    [KeyCodes.BURN_CODE]: [1, ClickType.NODE],
    [KeyCodes.RAGE_CODE]: [0, ClickType.BLANK],
    [KeyCodes.CANNON_CODE]: [2, ClickType.NODE],
    [KeyCodes.PUMP_CODE]: [1, ClickType.NODE],
};

interface EventParameters {
    [index: string]: [number, ClickType]; // Adjust the types according to what Event constructor expects
}

export const EVENTS: EventParameters = {
    [EventCodes.CANNON_SHOT_CODE]: [2, ClickType.NODE],
    [EventCodes.PUMP_DRAIN_CODE]: [1, ClickType.NODE],
    [EventCodes.STANDARD_LEFT_CLICK]: [1, ClickType.EDGE],
    [EventCodes.STANDARD_RIGHT_CLICK]: [1, ClickType.EDGE],
};
